contributors/design-proposals/resource-class.md

```diff
@@ -0,0 +1,363 @@
+// +nonNamespaced=true
+// +genclient=true
+
+type ResourceClass struct {
+        metav1.TypeMeta
+        metav1.ObjectMeta
+        // Spec defines resources required
+        Spec ResourceClassSpec
+        // +optional
+        Status ResourceClassStatus
+}
+// Spec defines resources required
+type ResourceClassSpec struct {
+        // Resource Selector selects resources
+        ResourceSelector []ResourcePropertySelector
+}
+
+// A null or empty selector matches no resources
+type ResourcePropertySelector struct {
+        // A list of resource/device selector requirements. ANDed from each ResourceSelectorRequirement
+        MatchExpressions []ResourceSelectorRequirement
+}
+
+// A resource selector requirement is a selector that contains values, a key, and an operator
+// that relates the key and values
+type ResourceSelectorRequirement struct {
+        // The label key that the selector applies to
+        // +patchMergeKey=key
+        // +patchStrategy=merge
+        Key string
+        // +optional
+        Values []string
+        // operator
+        Operator ResourceSelectorOperator
+}
+type ResourceSelectorOperator string
+
+const (
+        ResourceSelectorOpIn           ResourceSelectorOperator = "In"
+        ResourceSelectorOpEq           ResourceSelectorOperator = "Eq"
+        ResourceSelectorOpNotIn        ResourceSelectorOperator = "NotIn"
+        ResourceSelectorOpExists       ResourceSelectorOperator = "Exists"
+        ResourceSelectorOpDoesNotExist ResourceSelectorOperator = "DoesNotExist"
+        ResourceSelectorOpGt           ResourceSelectorOperator = "Gt"
+        ResourceSelectorOpGtEq         ResourceSelectorOperator = "GtEq"
+        ResourceSelectorOpLt           ResourceSelectorOperator = "Lt"
+        ResourceSelectorOpLtEq         ResourceSelectorOperator = "LtEq"
+)
+```
+### ResourceClassStatus
+```golang
+type ResourceClassStatus struct {
+        Allocatable resources.Quantity
+        Requested   resources.Quantity
+}
+```
+ResourceClass status is updated by the scheduler at:
+1. New *Resource Class* object creation.
+2. Node addition to the cluster.
+3. Node removal from the cluster.
+4. Pod creation if pod requests a resource class.
+5. Pod deletion if pod was consuming resource class.
+
+`ResourceClassStatus` serves the following two purposes:
+* Scheduler predicates evaluation while pod creation. For details, please refer
+  further sections
+* User can view the current usage/availability details about the resource class
+  using kubectl.
+
+### User story
+The administrator has deployed device plugins to support hardware present in the
+cluster. Device plugins, running on nodes, will update node status indicating
+the presence of this hardware. To offer this hardware to applications deployed
+on kubernetes in a portable way, the administrator creates a number of resource
+classes to represent that hardware. These resource classes will include metadata
+about the devices as selection criteria.
```

Comment:
Can you give an example how resource classes know which devices are selectable? AFAIK, as mentioned above, the administrator creates a resource class without any devices info or node info, who is responsible to do the mapping (ResourceClass --> Devices) work?

Reply:
Scheduler will watch apiserver for resource class objects creation and then scheduler already has nodeinfo in the cache. For each kind of device, a deviceinfo structure will be instantiated and list of device info will reside in nodeinfo. At each resource class creation, scheduler will iterate over the deviceinfo list and if resourceclass matches the device, deviceinfo's list of resourceclass references is updated.

For more detalis take a look at PoC: https://github.com/vikaschoudhary16/kubernetes/pull/3/files#diff-558cb8bde14dca10a3151bfc222a3aae